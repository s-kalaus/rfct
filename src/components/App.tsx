import React, { useState } from 'react';
import { EntityHighlighter } from './EntityHighlighter';
import { IEntity } from '~/interfaces/entity';
import { IStyleSheet } from '~/interfaces/style-sheet';

const App = () => {
  const [text, setText] = useState(
    'Venture first mover advantage learning curve market ecosystem funding stealth disruptive social proof scrum project growth hacking niche market user experience graphical user interface.'
  );
  const [entities, setEntities] = useState<IEntity[]>([
    { selectionStart: 160, selectionEnd: 184, label: 'very important' },
    { selectionStart: 144, selectionEnd: 159, label: 'very important' },
    { selectionStart: 62, selectionEnd: 69, label: 'important' },
    { selectionStart: 116, selectionEnd: 130, label: 'nonsense' },
    { selectionStart: 8, selectionEnd: 29, label: 'nonsense' },
  ]);

  return (
    <div style={styles.app}>
      <EntityHighlighter
        text={text}
        entities={entities}
        onTextChange={setText}
        onEntitiesChange={setEntities}
      />
    </div>
  );
};

const styles: IStyleSheet = {
  app: {
    textAlign: 'center',
    padding: '2em',
    width: '60%',
    margin: '0 auto',
    color: 'black',
    maxWidth: '750px',
    minWidth: '250px',
  },
  logo: {
    height: '5vmin',
  },
};

export default App;
