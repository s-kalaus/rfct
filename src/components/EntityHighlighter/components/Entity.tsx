import React, { useState } from 'react';
import { IEntity } from '~/interfaces/entity';
import { ENTITY_COLORS } from '../consts/entity-colors';
import { IStyleSheet } from '~/interfaces/style-sheet';
import { hashString } from '~/utils/hash-string';

export interface IProps {
  text: string;
  entity: IEntity;
}

export const Entity = ({ text, entity }: IProps) => {
  const [start] = useState(text.substr(0, entity.selectionStart));
  const [value] = useState(
    text.substr(
      entity.selectionStart,
      entity.selectionEnd - entity.selectionStart
    )
  );
  const [end] = useState(text.substr(entity.selectionEnd));
  const [backgroundColor] = useState(
    ENTITY_COLORS[hashString(entity.label) % ENTITY_COLORS.length].bg
  );

  return (
    <div style={styles.item}>
      {start}
      <span style={{ ...styles.highlight, backgroundColor }}>{value}</span>
      {end}
    </div>
  );
};

const styles: IStyleSheet = {
  item: {
    textAlign: 'left',
    position: 'absolute',
    top: '6px',
    left: '6px',
    right: '6px',
    color: 'transparent',
    pointerEvents: 'none',
    padding: '0',
    whiteSpace: 'pre-wrap',
    wordBreak: 'break-word',
    fontFamily:
      'source-code-pro, Menlo, Monaco, Consolas, "Courier New", monospace',
    fontSize: '14px',
  },
  highlight: {
    opacity: 0.3,
  },
};
