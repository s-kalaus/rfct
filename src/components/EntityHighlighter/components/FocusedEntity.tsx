import React from 'react';
import { IEntity } from '~/interfaces/entity';
import { IStyleSheet } from '~/interfaces/style-sheet';

export interface IProps {
  text: string;
  entity: IEntity;
  onDelete: (entity: IEntity) => void;
}

export const FocusedEntity = ({ text, entity, onDelete }: IProps) => (
  <div style={styles.frame}>
    {text.substring(entity.selectionStart, entity.selectionEnd)}
    {entity.label && ` (${entity.label})`}
    <button
      style={styles.button}
      onClick={() => onDelete(entity)}
      data-testid="button"
    >
      <span role="img" aria-label="Delete">
        🗑️
      </span>
    </button>
  </div>
);

const styles: IStyleSheet = {
  frame: { wordBreak: 'break-word' },
  button: {
    border: '0 none',
    cursor: 'pointer',
    backgroundColor: 'transparent',
  },
  highlight: {
    opacity: 0.3,
  },
};
