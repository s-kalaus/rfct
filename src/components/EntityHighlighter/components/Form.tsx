import React from 'react';

export interface IProps {
  disabled: boolean;
  label: string;
  onSubmit: (label: string) => void;
  onLabelChange: (label: string) => void;
}

export const Form = ({ label, disabled, onSubmit, onLabelChange }: IProps) => (
  <div>
    <input
      data-testid="input"
      type="text"
      placeholder="Entity label"
      value={label}
      onChange={(event) => onLabelChange(event.target.value)}
      disabled={disabled}
    />
    <button
      onClick={() => onSubmit(label)}
      disabled={disabled}
      data-testid="button"
    >
      Add entity for selection
    </button>
  </div>
);
