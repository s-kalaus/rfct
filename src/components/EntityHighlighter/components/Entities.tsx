import React from 'react';
import { IEntity } from '~/interfaces/entity';
import { Entity } from './Entity';

export interface IProps {
  text: string;
  entities: IEntity[];
}

export const Entities = ({ text, entities }: IProps) => (
  <>
    {entities.map((entity) => (
      <Entity text={text} entity={entity} key={JSON.stringify(entity)} />
    ))}
  </>
);
