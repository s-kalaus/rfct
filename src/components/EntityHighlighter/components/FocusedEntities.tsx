import React from 'react';
import { IEntity } from '~/interfaces/entity';
import { FocusedEntity } from './FocusedEntity';

export interface IProps {
  text: string;
  entities: IEntity[];
  onDelete: (entity: IEntity) => void;
}

export const FocusedEntities = ({ entities, text, onDelete }: IProps) => (
  <>
    {entities.map((entity) => (
      <FocusedEntity
        text={text}
        entity={entity}
        key={JSON.stringify(entity)}
        onDelete={onDelete}
      />
    ))}
  </>
);
