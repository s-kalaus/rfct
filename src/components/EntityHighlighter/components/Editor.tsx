import React from 'react';
import { IEntity } from '~/interfaces/entity';
import { Entities } from './Entities';
import { IStyleSheet } from '~/interfaces/style-sheet';

export interface IProps {
  text: string;
  entities: IEntity[];
  onTextChange: (text: string) => void;
  onInputRef: (node: HTMLTextAreaElement) => void;
}

export const Editor = ({
  text,
  onTextChange,
  entities,
  onInputRef,
}: IProps) => (
  <div style={styles.wrapper}>
    <textarea
      style={styles.input}
      ref={onInputRef}
      onChange={(event) => onTextChange(event.target.value)}
      value={text}
      rows={10}
      data-testid="textarea"
    />
    <Entities entities={entities} text={text} />
  </div>
);

const styles: IStyleSheet = {
  wrapper: {
    position: 'relative',
  },
  input: {
    fontFamily:
      'source-code-pro, Menlo, Monaco, Consolas, "Courier New", monospace',
    fontSize: '14px',
    background: 'none',
    border: '1px solid',
    width: '100%',
    resize: 'none',
    padding: '5px',
    whiteSpace: 'pre-wrap',
    boxSizing: 'border-box',
    wordBreak: 'break-word',
  },
};
