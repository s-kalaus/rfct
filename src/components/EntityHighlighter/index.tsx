import React, { MutableRefObject } from 'react';
import { IEntity } from '~/interfaces/entity';
import { useEntityHighlighter } from './hooks/useEntityHighlighter';
import { Form } from './components/Form';
import { FocusedEntities } from './components/FocusedEntities';
import { Editor } from './components/Editor';

export interface IProps {
  text: string;
  entities: IEntity[];
  onTextChange: (text: string) => void;
  onEntitiesChange: (entities: IEntity[]) => void;
}

export const EntityHighlighter = (props: IProps) => {
  const { text, entities } = props;
  const {
    focusedEntities,
    handleSelectionChange,
    hasZeroSelection,
    deleteEntity,
    addEntity,
    handleTextChange,
    selection,
    inputNode,
  } = useEntityHighlighter(props);

  return (
    <>
      <Editor
        onInputRef={setInputRef(inputNode)}
        text={text}
        entities={entities}
        onTextChange={handleTextChange}
      />
      <Form
        label={selection.label}
        disabled={hasZeroSelection}
        onSubmit={addEntity}
        onLabelChange={handleSelectionChange}
      />
      <FocusedEntities
        text={text}
        entities={focusedEntities}
        onDelete={deleteEntity}
      />
    </>
  );
};

const setInputRef =
  (inputNode: MutableRefObject<HTMLTextAreaElement | null>) =>
  (node: HTMLTextAreaElement) => {
    if (node) {
      inputNode.current = node;
    }
  };
