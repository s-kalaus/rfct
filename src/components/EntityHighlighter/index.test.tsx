import React from 'react';
import { render } from '@testing-library/react';
import { EntityHighlighter } from './index';
import { IProps as IEditorProps } from './components/Editor';
import { IProps as IFormProps } from './components/Form';
import { IProps as IFocusedEntitiesProps } from './components/FocusedEntities';

let mockRef: any;

jest.mock('./components/Editor', () => ({
  Editor: (props: IEditorProps) => {
    const { onInputRef } = props;
    onInputRef(mockRef);
    return `mock-editor: ${JSON.stringify({ ...props, mockRef })}`;
  },
}));
jest.mock('./components/Form', () => ({
  Form: (props: IFormProps) => `mock-form: ${JSON.stringify(props)}`,
}));
jest.mock('./components/FocusedEntities', () => ({
  FocusedEntities: (props: IFocusedEntitiesProps) =>
    `mock-focused-entities: ${JSON.stringify(props)}`,
}));
jest.mock('./hooks/useEntityHighlighter', () => ({
  useEntityHighlighter: () => ({
    focusedEntities: [],
    handleSelectionChange: jest.fn(),
    hasZeroSelection: true,
    deleteEntity: jest.fn(),
    addEntity: jest.fn(),
    handleTextChange: jest.fn(),
    selection: {
      selectionStart: 0,
      selectionEnd: 0,
      label: '',
    },
    inputNode: { current: null },
  }),
}));

describe('index', () => {
  const entities = [{ selectionStart: 0, selectionEnd: 5, label: 'Test' }];
  const text = 'Hello World';

  beforeEach(() => {
    mockRef = {
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
    };
  });

  test('renders component', () => {
    const { container } = render(
      <EntityHighlighter
        entities={entities}
        text={text}
        onTextChange={jest.fn()}
        onEntitiesChange={jest.fn()}
      />
    );
    expect(container).toMatchSnapshot();
  });

  test('renders component: no ref', () => {
    mockRef = null;
    const { container } = render(
      <EntityHighlighter
        entities={entities}
        text={text}
        onTextChange={jest.fn()}
        onEntitiesChange={jest.fn()}
      />
    );
    expect(container).toMatchSnapshot();
  });
});
