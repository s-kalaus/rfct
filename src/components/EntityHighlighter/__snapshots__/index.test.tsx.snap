// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index renders component 1`] = `
<div>
  mock-editor: {"text":"Hello World","entities":[{"selectionStart":0,"selectionEnd":5,"label":"Test"}],"mockRef":{}}
  mock-form: {"label":"","disabled":true}
  mock-focused-entities: {"text":"Hello World","entities":[]}
</div>
`;

exports[`index renders component: no ref 1`] = `
<div>
  mock-editor: {"text":"Hello World","entities":[{"selectionStart":0,"selectionEnd":5,"label":"Test"}],"mockRef":null}
  mock-form: {"label":"","disabled":true}
  mock-focused-entities: {"text":"Hello World","entities":[]}
</div>
`;
